File: app/src/main/java/com/codemybrainsout/imageviewer/view/single/SinglePresenter.java
package com.codemybrainsout.imageviewer.view.single;

import com.codemybrainsout.imageviewer.api.CollectionService;
import com.codemybrainsout.imageviewer.model.Collection;
import com.codemybrainsout.imageviewer.model.Photo;
import com.codemybrainsout.imageviewer.model.User;

import java.util.List;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

/**
 * Created by ahulr on 11-06-2017.
 */

public class SinglePresenter implements SingleContract.Presenter {

    private Collection collection;
    public CollectionService mCollectionService;
    public SingleContract.View mView;

    public SinglePresenter(CollectionService collectionService, SingleContract.View view, Collection collection) {
        this.mCollectionService = collectionService;
        this.mView = view;
        this.collection = collection;
    }

    @Override
    public void openUser(User user) {
        mView.showUser(user);
    }

    @Override
    public void openPhoto(Photo photo) {
        mView.showPhoto(photo);
    }

    @Override
    public void loadCollection(Collection collection) {
        mView.showCollectionDetails(collection);
    }

    @Override
    public void loadPhotos(int page, int limit) {

        mView.showLoading();
        mCollectionService.getSingleCollection(collection.getId(), page, limit)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<List<Photo>>() {
                    @Override
                    public void onSubscribe(@NonNull Disposable d) {

                    }

                    @Override
                    public void onNext(@NonNull List<Photo> photos) {
                        mView.refreshPhotos(photos);
                        mView.hideLoading();
                    }

                    @Override
                    public void onError(@NonNull Throwable e) {
                        e.printStackTrace();
                        mView.showError(e.getMessage());
                        mView.hideLoading();
                    }

                    @Override
                    public void onComplete() {

                    }
                });


    }

    @Override
    public void loadMorePhotos(int page, int limit) {

        mView.showLoading();
        mCollectionService.getSingleCollection(collection.getId(), page, limit)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<List<Photo>>() {
                    @Override
                    public void onSubscribe(@NonNull Disposable d) {

                    }

                    @Override
                    public void onNext(@NonNull List<Photo> photos) {
                        mView.addPhotos(photos);
                        mView.hideLoading();
                    }

                    @Override
                    public void onError(@NonNull Throwable e) {
                        e.printStackTrace();
                        mView.showError(e.getMessage());
                        mView.hideLoading();
                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }
}

File: app/src/main/java/com/codemybrainsout/imageviewer/view/single/SinglePresenter.kt
package com.codemybrainsout.imageviewer.view.single

import com.codemybrainsout.imageviewer.api.CollectionService
import com.codemybrainsout.imageviewer.model.Collection
import com.codemybrainsout.imageviewer.model.Photo
import com.codemybrainsout.imageviewer.model.User
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by ahulr on 11-06-2017.
 */
class SinglePresenter @Inject constructor(
    private val collectionService: CollectionService,
    val view: SingleContract.View,
) : SingleContract.Presenter {

    override fun openUser(user: User?) {
        view.showUser(user)
    }

    override fun openPhoto(photo: Photo?) {
        view.showPhoto(photo)
    }

    override fun loadCollection(collection: Collection?) {
        view.showCollectionDetails(collection)
    }

    override fun loadPhotos(collectionId: Int, page: Int, limit: Int) {
        view.showLoading()
        collectionService.getSingleCollection(collectionId, page, limit)
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({
                view.refreshPhotos(it)
                view.hideLoading()
            }, {
                it.printStackTrace()
                view.showError(it.message)
                view.hideLoading()
            })
    }

    override fun loadMorePhotos(collectionId: Int, page: Int, limit: Int) {
        view.showLoading()
        collectionService.getSingleCollection(collectionId, page, limit)
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                {
                    view.addPhotos(it)
                    view.hideLoading()
                },
                {
                    it.printStackTrace()
                    view.showError(it.message)
                    view.hideLoading()
                }
            )
    }
}